### DISCUSSION TITLE
What's the best way to query the cell size?

### DISCUSSION THREAD

**Discussion by gzagatti** [kovidgoyal/kitty#5287] [Q&A]
I need to query the terminal about its cell size in pixel. 

I am aware of the [control sequence `CSI 1 6 t`](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h4-Functions-using-CSI-_-ordered-by-the-final-character-lparen-s-rparen:CSI-Ps;Ps;Ps-t.1EB0), but I am not able to capture it's output. I have tried the following command:

```sh
> printf '\e[16t'

> ;19;10t
```

The command will print an empty line and then return the cell size on stdin. 

I could not save the result of the escape sequence. Is there a way to do so? Or is there a kitten that will return this information? 

For instance, `kitty +kitten icat --print-window-size` will return the window size. I have noticed that `icat` gets the window size from an object defined in `kitty.utils` that also contains the cell size information. 

I wanted to avoid writing a custom kitty just to get this information if possible.

**Comment by page-down** [comment]
You can use `read` and disable echo to read the returned results.
For example, in bash:

```bash
IFS=';' read -rs -d t -p $'\e[16t' -a CELL_SIZE
echo "height: ${CELL_SIZE[1]}, width: ${CELL_SIZE[2]}"
```

You can also write a kitten for this.

**Comment by kovidgoyal** [comment]
https://sw.kovidgoyal.net/kitty/graphics-protocol/#getting-the-window-size

**Comment by kovidgoyal** [comment]
https://sw.kovidgoyal.net/kitty/graphics-protocol/#getting-the-window-size

**Comment by gzagatti** [comment]
Thanks! The read method works very well on the terminal. 

Now, I'm figuring out how to adapt the Python code you shared for Lua. As my objective is to work on a Neovim plugin for images.

[Hologram.nvim](https://github.com/edluffy/hologram.nvim/blob/main/lua/hologram/terminal.lua#L4) uses [libuv](https://github.com/luvit/luv/blob/master/docs.md#uv_stream_t--stream-handle) to interact with the tty. However, I am not yet sure how to make the appropriate call from to get the cell size.

**Comment by gzagatti** [comment]
Ok, so I got a lua script that can output the cell size:

```lua
ffi = require'ffi'
ffi.cdef[[
int ioctl(int __fd, unsigned long int __request, ...);
typedef struct winsize {
  unsigned short ws_row;
  unsigned short ws_col;
  unsigned short ws_xpixel;
  unsigned short ws_ypixel;
};
]]
sz = ffi.new('struct winsize')
ffi.C.ioctl(0, 21523, sz)
sz.ws_ypixel / sz.ws_row
```

The script basically wraps the C implementation from your link above. I am not sure if this is the most elegant way of doing this. Perhaps going the `libuv` way would be better since it's already used in `hologram.nvim`.
